### PR TITLE
mrc-4390: Add first part of child model

### PR DESCRIPTION
**Pull request by r-ash**
This PR will
* Add structure for paediatric/child model, which I hope will be extensible to other types of models

### CODE REVIEW

File: inst/include/hiv_demographic_projection.hpp [hunk 1]
```diff
@@ -152,18 +172,22 @@ void run_hiv_and_art_stratified_deaths_and_migration(
...
-                                                 intermediate);
-  internal::run_hiv_and_art_stratified_deaths_and_migration<S>(time_step, pars, state_curr,
-                                                               state_next, intermediate);
+                                        const Parameters<ModelVariant, real_type> &pars,
+                                        const State<ModelVariant, real_type> &state_curr,
+                                        State<ModelVariant, real_type> &state_next,
+                                        internal::IntermediateData<ModelVariant, real_type> &intermediate) {
+  internal::run_hiv_ageing_and_mortality<ModelVariant>(time_step, pars, state_curr, state_next,
+                                                       intermediate);
+  if (ModelVariant::run_child_model) {
```
Comment (r-ash) [inline inst/include/hiv_demographic_projection.hpp:183]: if constexpr and remove if in run_age_15_entrants above

File: inst/include/child_model_simulation.hpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+}
+}
+
+template<typename ModelVariant, typename real_type>
+void run_child_model_simulation(int time_step,
+                                const Parameters<ModelVariant, real_type> &pars,
+                                const State<ModelVariant, real_type> &state_curr,
+                                State<ModelVariant, real_type> &state_next,
+                                internal::IntermediateData<ModelVariant, real_type> &intermediate) {
+  internal::run_hiv_child_infections(time_step, pars, state_curr, state_next, intermediate);
```
Comment (r-ash) [inline inst/include/child_model_simulation.hpp:45]: Add a static assert here https://github.com/mrc-ide/dust/blob/master/inst/include/dust/random/binomial.hpp#L288

File: inst/include/state_saver.hpp [hunk 3]
```diff
@@ -83,35 +118,42 @@ class StateSaver {
   }
 
 
-  void save_state(const State<S, real_type> &state, int current_year) {
+  void save_state(const State<ModelVariant, real_type> &state, int current_year) {
```
Comment (r-ash) [inline inst/include/state_saver.hpp:121]: Split this state saver on the ModelVariant too. And have this call the save_state from the base and child StateSaver objects

File: inst/include/model_output.hpp [hunk 4]
```diff
@@ -10,68 +10,85 @@
...
-  std::copy_n(state.h_art_initiation.data(), state.h_art_initiation.size(), REAL(r_h_art_initiation));
-  std::copy_n(state.p_hiv_deaths.data(), state.p_hiv_deaths.size(), REAL(r_p_hiv_deaths));
+  if constexpr (ModelVariant::run_child_model) {
+    constexpr auto base = ss.base;
+    constexpr auto children = ss.children;
+    Rcpp::NumericVector r_hc_hiv_pop(base.hDS * children.hTM * base.pAG * base.NS * output_years);
+    r_hc_hiv_pop.attr("dim") = Rcpp::NumericVector::create(base.hDS, children.hTM, base.pAG, base.NS, output_years);
+    std::copy_n(state.children.hc_hiv_pop.data(), state.children.hc_hiv_pop.size(), REAL(r_hc_hiv_pop));
+    ret.push_back(r_hc_hiv_pop, "hc_hiv_pop");
+  }
```
Comment (r-ash) [inline inst/include/model_output.hpp:91]: Should we return here a list of lists? `base`, `child` etc. and should the input data also be split like this?

File: scripts/update_standlone_data [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+parameters <- readRDS(testthat::test_path("testdata/projection_parameters_adult.rds"))
+
+frogger:::serialize_r_to_tensor(demp$basepop[, , 1], "inst/fit_model/data/basepop")
+frogger:::serialize_r_to_tensor(demp$Sx, "inst/fit_model/data/survival")
+frogger:::serialize_r_to_tensor(demp$netmigr, "inst/fit_model/data/net_migration")
+frogger:::serialize_r_to_tensor(demp$asfr, "inst/fit_model/data/age_sex_fertility_ratio")
+frogger:::serialize_r_to_tensor(demp$births_sex_prop, "inst/fit_model/data/births_sex_prop")
+frogger:::serialize_r_to_tensor(parameters$incidinput, "inst/fit_model/data/incidence_rate")
+frogger:::serialize_r_to_tensor(parameters$incrr_age, "inst/fit_model/data/incidence_rate_relative_risk_age")
+frogger:::serialize_r_to_tensor(parameters$incrr_sex, "inst/fit_model/data/incidence_rate_relative_risk_sex")
```
Comment (r-ash) [inline scripts/update_standlone_data:15]: Document and tidy this up. Why would someone run this script?

Comment (r-ash) [reply scripts/update_standlone_data:15]: Could run this on CI and see if the files have changed

File: src/frogger.cpp [hunk 6]
```diff
@@ -78,22 +67,24 @@ Rcpp::List run_base_model(const Rcpp::List data,
...
+  } else if (hiv_age_stratification == "full") {
+    constexpr auto ss = leapfrog::StateSpace<leapfrog::BaseModelFullAgeStratification>();
+    ret = fit_model<leapfrog::BaseModelFullAgeStratification>(ss, data, proj_years, hiv_steps, save_steps);
   } else {
     // We've already validated that hiv_age_stratification is one of "full" or "coarse"
-    constexpr auto ss = leapfrog::StateSpace<leapfrog::HivAgeStratification::coarse>();
-    ret = fit_model<leapfrog::HivAgeStratification::coarse>(ss, data,
-                                                            proj_years, hiv_steps, save_steps);
+    constexpr auto ss = leapfrog::StateSpace<leapfrog::BaseModelCoarseAgeStratification>();
+    ret = fit_model<leapfrog::BaseModelCoarseAgeStratification>(ss, data, proj_years, hiv_steps, save_steps);
```
[the remark this answers is not in the document]
Comment (r-ash) [reply src/frogger.cpp:87]: Instead think about passing in the name of the model variant and then switch on that. And handle the traits to model variant selection in R so we can have nicer error messages

File: scripts/update_standlone_data (hunk 5 above)
Comment (r-ash) [reply scripts/update_standlone_data:15]: Made a ticket for this https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4503/Add-test-which-checks-the-standalone-test-data-is-not-out-of-date

File: inst/include/model_output.hpp (hunk 4 above)
Comment (r-ash) [reply inst/include/model_output.hpp:91]: Will address this in another PR immediately after this one

File: R/util.R [hunk 7]
```diff
@@ -16,7 +16,8 @@ serialize_r_to_tensor <- function(data, path) {
   if (is.integer(data)) {
     type <- "int"
   } else if (is.logical(data)) {
-     ## using as.integer() removes any dimensions. *1 to preserve shape and convert to int
+    ## using as.integer() removes any dimensions.
+    ## *1 to preserve shape and convert to int
     data <- data * 1
```
Comment (richfitz) [inline R/util.R:21]: fyi, also:

```
storage.mode(data) <- "integer"
```

(which will make this actually integer rather than numeric)

**Review by richfitz — APPROVED** [review]

File: inst/include/model_output.hpp (hunk 4 above)
Comment (r-ash) [reply inst/include/model_output.hpp:91]: Actually, made a ticket for this https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4516/Split-the-input-and-output-data-into-lists-of-lists-split-by-model-type

**Pull request merged**